### DISCUSSION TITLE
getProviders doesn't work on getServerSideProps

### DISCUSSION THREAD

**Discussion by akhfaern** [nextauthjs/next-auth#3535] [Help]
### Description 🐜

when calling getProviders on getServerSideProps it generates an error and returns null

### Is this a bug in your own project?

Yes

### How to reproduce ☕️

```js
export async function getServerSideProps(context) {
  const providers = await getProviders();

  return {
    props: {
      providers,
    },
  };
}
```

### Screenshots / Logs 📽

```sh
[next-auth][error][CLIENT_FETCH_ERROR] 
https://next-auth.js.org/errors#client_fetch_error invalid json response body at http://localhost:3000/api/auth/providers reason: Unexpected token < in JSON at position 0 {
  error: {
    message: 'invalid json response body at http://localhost:3000/api/auth/providers reason: Unexpected token < in JSON at position 0',
    stack: 'FetchError: invalid json response body at http://localhost:3000/api/auth/providers reason: Unexpected token < in JSON at position 0\n' +
      '    at /Users/muratcemyalin/ReactProjects/instagram-clone/node_modules/node-fetch/lib/index.js:272:32\n' +
      '    at processTicksAndRejections (internal/process/task_queues.js:95:5)',
    name: 'FetchError'
  },
  path: 'providers',
  message: 'invalid json response body at http://localhost:3000/api/auth/providers reason: Unexpected token < in JSON at position 0'
}
```

### Environment 🖥

  System:
    OS: macOS 11.6
    CPU: (8) x64 Apple M1
    Memory: 42.23 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.18.1 - /usr/local/bin/node
    npm: 6.14.15 - /usr/local/bin/npm
  Browsers:
    Chrome: 96.0.4664.110
    Firefox: 93.0
    Safari: 15.0
  npmPackages:
    next: 12.0.7 => 12.0.7 
    next-auth: ^4.0.6 => 4.0.6 
    react: 17.0.2 => 17.0.2 

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
I could not reproduce. Please provide a minimal reproduction.

**Comment by akhfaern** [comment]
i've found this question at stackoverflow who encountered same problem with me. 

https://stackoverflow.com/questions/70050759/trying-to-configure-next-auth-signin-page-but-having-a-problem-with-getprovider

you can see the minimal reproduction there

but if i go to http://localhost:3000/api/auth/providers it returns

` {"google":{"id":"google","name":"Google","type":"oauth","signinUrl":"http://localhost:3000/api/auth/signin/google","callbackUrl":"http://localhost:3000/api/auth/callback/google"}}`

 which is normal so i think there is a problem with async await in getServerSideProps

and also it works fine if i move getProviders to client side like:

```js
const [providers, setProviders] = useState(null);

  useEffect(() => {
    (async () => {
      const res = await getProviders();
      setProviders(res);
    })();
  }, []);
```

**Reply by nandhahacker1** [reply]
it's work

**Comment by balazsorban44** [comment]
Since you opened this issue, I expect a reproduction from you, not someone else's from stack overflow. Please take the time to report an issue if you would like it resolved as quickly as possible. It helps both of us.

On the stack overflow link it was also suggested that you double check your `NEXTAUTH_URL` value. I recommend the same.

**Comment by akhfaern** [comment]
here is my total 10 lines of code to reproduction:
https://github.com/akhfaern/nextauth-bug-report
requested url: http://localhost:3000/auth/signin

of course i checked NEXTAUTH_URL . as you didnt read my comments i said 
error: {
    message: 'invalid json response body at http://localhost:3000/api/auth/providers reason: Unexpected token < in JSON at position 0',

but when i opened in new tab http://localhost:3000/api/auth/providers it just works fine and returns json data.
also when i moved getProviders to component body It just works fine too! it just doesnt work in getServerSideProps. but in nextauth documents it says it has to work. also there are lots of examples on internet that shows the same code working i guess some update broke it.

**Comment by imphillipzissou** [comment]
Hey folks, I was running into this issue earlier today on a new next/react project. I was using the same versions as noted above. Ultimately, I ended up stumbling upon a blog [article](https://balazsorban.com/blog/next-auth-v4#getserversession) about v4 ServerSideProps. I modified mine to use `getServerSession` instead of `getSession` and it seems to be working fine now.

```
import { GetServerSidePropsContext } from 'next'
import { authOptions } from "../api/auth/[...nextauth]"

export async function getServerSideProps(context:GetServerSidePropsContext) {
  const session = await getServerSession(context, authOptions)
  
  console.log(session);
  
  return {
    props: {
      session,
    },
  }
}
```
Best!

edit: realized this was for session and not providers; I was having the issue with both sessions and providers.

dbl edit: found that my issue was related to the reverse proxy and the forward auth. could have sworn I checked this too.

**Comment by tamvo22** [comment]
Hello @akhfaern,

From your repo, I found that it's missing the following. After I add those missing items, the Google provider button shows up fine for me. You can clone the example repo and start from there if you're still having issues. 
[https://github.com/nextauthjs/next-auth-example](https://github.com/nextauthjs/next-auth-example)

- _app is missing SessionProvider
- [...nextauth].js is missing session jwt strategy and secret

_app:
```
import { SessionProvider } from 'next-auth/react';
import '../styles/globals.css';

function MyApp({ Component, pageProps }) {
  return (
    <SessionProvider session={pageProps.session}>
      <Component {...pageProps} />
    </SessionProvider>
  );
}

export default MyApp;
```
[...nextauth].js
```
import NextAuth from 'next-auth';
import GoogleProvider from 'next-auth/providers/google';

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
    // ...add more providers here
  ],
  secret: process.env.SECRET,

  session: {
    strategy: 'jwt',
  },
  pages: {
    signIn: '/auth/signin',
    error: '/auth/signin',
  },
});
```

**Comment by akhfaern** [comment]
hello @tamvo22, i've updated the repo in according to your code samples. still not working.. still the same error. and your sample repo does not helping because it has no custom signin page and what i am stuck is making a custom sign in page with getProviders function. actully i made it work without SSR but i want to do it in getServerSideProps which doesnt work

**Comment by dzerium** [comment]
Experienced this same thing too. I tried containerizing the app and running it through Docker and it's all working as expected. (weird I know)

**Comment by MohamedHakem** [comment]
Hi @akhfaern,  in case you still having trouble with this issue, I've made an example app and I'll list the important parts here, and you can follow, check your project and see where's the missing part in your code. 



**- Make a file called [...nextauth].js on /pages/api/auth/[...nextauth].js**
```
import NextAuth from "next-auth";
import GoogleProvider from "next-auth/providers/google";

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
  pages: {
    signIn: "/auth/signin",
  },
  secret: process.env.SECRET,
});

```



**- Make a file called signin.js on /pages/auth/signin.js**

```
import { getProviders, signIn } from "next-auth/react";

// browser side,
export default function SignIn({ providers }) {
  // log the providers to see for yourself
  console.log("Providers: ", providers);

  return (
    <div className="flex flex-col items-center justify-center min-h-screen py-2 font-bold text-2xl">
        <p>This is the signin page!</p>
        
        {Object.values(providers).map((provider) => (
          
          <div key={provider.name}>
              <button className="flex w-auto mt-10 p-4 bg-blue-500 text-white h-auto rounded-lg focus:outline-none"
                onClick={() => signIn(provider.id, { callbackUrl: "/" })} >

                Log in with {provider.name}
              </button>
          </div>
        ))}
    </div>
  );
}

// server side
export async function getServerSideProps() {
      const providers = await getProviders();
      return {
            props: { providers },
      };
}
```




**- In your   _app.js file, make sure it's similar to this:** 

```
import "tailwindcss/tailwind.css";
import { SessionProvider } from "next-auth/react";

export default function MyApp({ Component, pageProps }) {
  return (
    <SessionProvider session={pageProps.session}>
      <Component {...pageProps} />
    </SessionProvider>
  );
}
```

**- In your   index.js file, make sure it's similar to this:** 

```
import Header from "../components/Header";
import { getSession, useSession, getProviders } from "next-auth/react";
import { useRouter } from "next/router";
import { useEffect } from "react";

export default function Home({ providers }) {
  // get the session to check if the user logged in or not
  const { data: session } = useSession();

  // check if there's a session, if not: redirect to /auth/signin page
  const router = useRouter();
  useEffect(() => {
    if (!session) return router.push("/auth/signin");
  }, []);

  return (
    <>
      {session ? (
        <>
          <Header />
          <p className="flex flex-col items-center justify-center min-h-screen py-2 font-bold text-2xl">
            You've logged in successfully!
          </p>
        </>
      ) : (
        <></> //Show nothing... Redirecting to /auth/signin (you can add a button here instead of useEffect)
      )}
    </>
  );
}

export async function getServerSideProps(context) {
  return {
    props: {
      //Why: To check if the user logged-in or not, get the session and pass it to Home component
      session: await getSession(context),
    },
  };
}
```

**The result:**
- When you go to the home page, for example localhost:3000 
It redirects you automatically to the signin page (localhost:3000/auth/signin.js)

![auto-redirect-on-homepage](https://user-images.githubusercontent.com/29775958/147616110-3fc32891-7d3d-4ed8-a354-8b63b326fd3c.JPG)

- After you login:

![logged-in-success](https://user-images.githubusercontent.com/29775958/147616146-e306c6bc-6f58-4966-9c4d-8476ffa78192.JPG)


I hope this helps.

**Reply by sdeagh** [reply]
I copied this example and I still get the problem. Works fine with yarn dev but breaks when you build and run it. What could I be missing?

**Comment by Joroze** [comment]
Still experiencing this issue.

**Comment by sdeagh** [comment]
I have this (or very similar) issue too. Everything works fine in development mode (yarn dev) but errors occur when building and then running the app (yarn build && yarn start). Calling getProviders() in getServerSideProps in a custom signIn page results in this error on the server: 

error: {
    message: 'request to http://localhost:3000/api/auth/providers failed, reason: connect ECONNREFUSED ::1:3000',
    stack: 'FetchError: request to http://localhost:3000/api/auth/providers failed, reason: connect ECONNREFUSED ::1:3000\n' +
    
  },

**Comment by itta611** [comment]
Hi @akhfaern

Did you create folder in /api/**auth/** ?

**Comment by magicspon** [comment]
I'm seeing the same issue:
example: https://github.com/magicspon/next-auth-custom-login-page-error

**Comment by mattkastner** [comment]
I was able to fix it by making a change in my `_app.js` file:
`export default function App({ Component, pageProps: { session, pageProps } }) { **SOME CODE ** }`
to
`export default function App({ Component, pageProps: { session, ...pageProps } }) { **SOME CODE ** }`

**Code Screenshot**
<img width="697" alt="Screen Shot 2022-05-10 at 10 37 16 AM" src="https://user-images.githubusercontent.com/32001990/167679043-5724ea88-f68f-4c12-90e7-66cad3ba064c.png">

-----------

I needed the providers for a custom "Sign in" page and my `getServerSideProps` in that file looked like this in the end:

**Code Screenshot**
<img width="441" alt="Screen Shot 2022-05-10 at 10 29 59 AM" src="https://user-images.githubusercontent.com/32001990/167677785-b2ca5b89-c686-4b3c-ac27-408fe8ea44a9.png">

-----------

Here is an **overview** of boths files and their location in my project.
<img width="1728" alt="Screen Shot 2022-05-10 at 10 35 17 AM" src="https://user-images.githubusercontent.com/32001990/167678740-6ab781f5-437f-4b68-966a-f11a28499f6a.png">

**Reply by josemsp** [reply]
**This answer is correct,** I am not expert but the main problem here is we need to understand that NEXT JS **has a router integrated** and we need to use them.
You need to redirect or use a router manager, because if we don't tell NEXT JS that what it will render is not in the current route and render another **getServerSideProps** component it won't work correctly.
This only applies when what we are trying to render is a view component.
I had the same problem and I solve it in a similar way, but redirecting to the "signin" page

> I was able to fix it by making a change in my `_app.js` file: `export default function App({ Component, pageProps: { session, pageProps } }) { **SOME CODE ** }` to `export default function App({ Component, pageProps: { session, ...pageProps } }) { **SOME CODE ** }`
> 
> **Code Screenshot** <img alt="Screen Shot 2022-05-10 at 10 37 16 AM" width="697" src="https://user-images.githubusercontent.com/32001990/167679043-5724ea88-f68f-4c12-90e7-66cad3ba064c.png">
> 
> I needed the providers for a custom "Sign in" page and my `getServerSideProps` in that file looked like this in the end:
> 
> **Code Screenshot** <img alt="Screen Shot 2022-05-10 at 10 29 59 AM" width="441" src="https://user-images.githubusercontent.com/32001990/167677785-b2ca5b89-c686-4b3c-ac27-408fe8ea44a9.png">
> 
> Here is an **overview** of boths files and their location in my project. <img alt="Screen Shot 2022-05-10 at 10 35 17 AM" width="1728" src="https://user-images.githubusercontent.com/32001990/167678740-6ab781f5-437f-4b68-966a-f11a28499f6a.png">

**Comment by SinhTNgoc** [comment]
Try to update node @v16!!!

**Comment by pekala** [comment]
For anyone hitting this - check if you don't have another app served on the same port - that was it for me.

**Reply by agrattan0820** [reply]
Can confirm, I had this too. Spent an entire day upgrading and reconfiguring `next-auth` hoping it would fix my issue. Ended up just being because I had [Local](https://localwp.com/) opened up on my computer and it was running on the same port 🙃

**Comment by tammAtPropertyMe** [comment]
I ran into this, and based on the advice in this thread and the note in nextjs docs that ["This produces an additional call, reducing performance."](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes), should you call getProviders() in your getServerSideProps at all? 

Maybe the answer could be not to do this, rather than tweaking things until it works.

An alternative if what you are after is "I want to list the providers in the next-auth config" would be to export the next-auth config [as per a lot of examples](https://next-auth.js.org/configuration/nextjs) and import it like this:

```
import { authOptions } from "../api/auth/[...nextauth]";

export async function getServerSideProps() {
  const providers = authOptions.providers;

  return {
    props: {
      providers,
    },
  };
}
```

I'd love input on whether this is a good solution, but it is A solution. ^_^

**Comment by rafael977** [comment]
For me `getProviders` works after I set `NODE_OPTIONS='--dns-result-order=ipv4first'`, i.e. `NODE_OPTIONS='--dns-result-order=ipv4first' next start` for the start script.
The reason seems to be nodejs after certain version, prefers ipv6 for dns lookup (not sure if it's the right explanation and doesn't explain my `next dev` work without setting the option). Anyway, when I run `next start` it resolves localhost to ::1 while the server is listening 0.0.0.0.
So if this being the root cause, other than setting NODE_OPTIONS, let next js server listen to ::1, i.e. `next start -H ::1` or simply set `NEXTAUTH_URL=http://127.0.0.1:3000` should work as well.

**Reply by formidabilus** [reply]
setting the NEXTAUTH_URL=http://127.0.0.1:3000 fixed it for me, thank you, I've wasted 3 days looking for this.

**Reply by ghost** [reply]
@balazsorban44 this answer here from @rafael977 seems to be correct and should give some good pointers to solving this. It also addresses a couple of other previously raised issues/discussions as far as I can tell, F. ex. https://github.com/nextauthjs/next-auth/issues/6171. The workaround listed as an answer in https://github.com/nextauthjs/next-auth/discussions/4870 does work, but it is not ideal. 

My use case when running in to this was runnning playwright tests on a production build with localhost:3000 as the NEXTAUTH_URL.

Happy give you a reproduction repo if you decide to look into this.

**Comment by Dujota** [comment]
Running into a strange issue/behavior with getProviders runnin on SSR. 

if i call it on a page.tsx (next server component) it returns a different list of providers vs calling this function on the client side 

It's literally giving me two different lists of providers, and the wierdest part is, even if i straight delete on provider from the authOptions Array, it still shows up on the server component call.

Is this being cached somewhere? I am a bit confused with this behavior since next-auth docs suggests that there is no special usage for this method

`const providers = await getProviders();`

vs 

```
const [providers, setProviders] = useState({});

  useEffect(() => {
    (async () => {
      const res = await getProviders();
      setProviders(res);
    })();
  }, []);
```

I had to jigsaw a solution as follows:

```const providers =
  authOptions()?.providers?.map((provider) => ({
    id: provider.id,
    name: provider.name,
    type: provider.type,
  })) || [];

**Reply by CyberClarence** [reply]
it's also returning me two different list of providers, something is cached for sure

**Reply by CyberClarence** [reply]
it call fetch under the hood so by default it will cache your first call  to /api/ auth/providers and that explain everything
<img width="985" alt="image" src="https://github.com/nextauthjs/next-auth/assets/136319693/a5a1a8e4-406a-4cf9-b2b3-af0919966725">
<img width="735" alt="image" src="https://github.com/nextauthjs/next-auth/assets/136319693/97f301fb-0d47-473b-bd27-1409e5c0535c">
